File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.py
def solution(str1, str2):
    s1 = [str1[i:i+2].lower()
          for i in range(len(str1)-1) if str1[i:i+2].isalpha()]
    s2 = [str2[i:i+2].lower()
          for i in range(len(str2)-1) if str2[i:i+2].isalpha()]

    intersection = set(s1) & set(s2)
    union = set(s1) | set(s2)

    if len(union) == 0:
        return 65536

    m_intersection = sum([min(s1.count(i), s2.count(i)) for i in intersection])
    m_union = sum([max(s1.count(u), s2.count(u)) for u in union])

    return int(65536 * (m_intersection / m_union))

File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int solution(String str1, String str2) {
        Map<String, Integer> map1 = createMap(str1);
        Map<String, Integer> map2 = createMap(str2);

        int intersection = intersection(map1, map2);
        int union = union(map1, map2);

        if (union == 0) {
            return 65536;
        }

        return (int) ((double) intersection / union * 65536);
    }

    public Map<String, Integer> createMap(String s) {
        Map<String, Integer> map = new HashMap<>();
        for (int i = 0; i < s.length() - 1; i++) {
            char c1 = s.charAt(i);
            char c2 = s.charAt(i + 1);
            if (!Character.isAlphabetic(c1) || !Character.isAlphabetic(c2)) {
                continue;
            }
            String sub = String.valueOf(c1).toUpperCase() + String.valueOf(c2).toUpperCase();
            map.put(sub, map.getOrDefault(sub, 0) + 1);
        }
        return map;
    }

    public int intersection(Map<String, Integer> map1, Map<String, Integer> map2) {
        int intersection = 0;
        for (Map.Entry<String, Integer> entry : map1.entrySet()) {
            String key = entry.getKey();
            intersection += Math.min(entry.getValue(), map2.getOrDefault(key, 0));
        }

        return intersection;
    }

    public int union(Map<String, Integer> map1, Map<String, Integer> map2) {
        int union = 0;
        for (Map.Entry<String, Integer> entry : map1.entrySet()) {
            String key = entry.getKey();
            union += Math.max(entry.getValue(), map2.getOrDefault(key, 0));
        }

        for (Map.Entry<String, Integer> entry : map2.entrySet()) {
            String key = entry.getKey();
            if (!map1.containsKey(key)) {
                union += entry.getValue();
            }
        }

        return union;
    }
}